services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/operations/UpdateItemOperation.java

```diff
@@ -141,6 +151,56 @@ public UpdateItemRequest generateRequest(TableSchema<T> tableSchema,
 
         return requestBuilder.build();
     }
+    
+    /**
+     * Method checks if a nested object parameter requires an update
+     * If so flattens out nested params separated by "_NESTED_ATTR_UPDATE_"
+     * this is consumed by @link EnhancedClientUtils to form the appropriate UpdateExpression
+     */
+    public Map<String, AttributeValue> transformItemToMapForUpdateExpression(Map<String, AttributeValue> itemToMap,
+                                                                             boolean ignoreNulls) {
+        
+        Map<String, AttributeValue> nestedAttributes = new HashMap<>();
+        
+        itemToMap.forEach((key, value) -> {
+            if (value.hasM()) {
+                nestedAttributes.put(key, value);
+            }
+        });
+        
+        if (!nestedAttributes.isEmpty()) {
+            Map<String, AttributeValue> itemToMapMutable = new HashMap<>(itemToMap);
+            nestedAttributes.forEach((key, value) -> {
+                itemToMapMutable.remove(key);
+                nestedItemToMap(itemToMapMutable, key, value, ignoreNulls);
+            });
+            return itemToMapMutable;
+        }
+        
+        return itemToMap;
+    }
+    
+    private Map<String, AttributeValue> nestedItemToMap(Map<String, AttributeValue> itemToMap,
+                                 String key,
+                                 AttributeValue attributeValue,
+                                 boolean ignoreNulls) {
+        attributeValue.m().forEach((mapKey, mapValue) -> {
+            String nestedAttributeKey = key + NESTED_OBJECT_UPDATE + mapKey;
+            if (attributeValueNonNullOrShouldWriteNull(ignoreNulls, mapValue)) {
+                if (mapValue.hasM()) {
+                    nestedItemToMap(itemToMap, nestedAttributeKey, mapValue,
+                                    ignoreNulls);
+                } else {
+                    itemToMap.put(nestedAttributeKey, mapValue);
+                }
+            }
+        });
+        return itemToMap;
+    }
+    
+    private boolean attributeValueNonNullOrShouldWriteNull(boolean ignoreNulls, AttributeValue attributeValue) {
```

Comment:
i think we will need to do some additional check here to make sure the intentional null are updated

example
### value in DDB
```json
{
  "id": "ignoreNullFalse",
  "nestedRecord": {
    "id":"nestedId",
    "nestedUpdateBehaviorAttribute": "FixedOverride",
    "nestedCounter": 9999
  },
  "version": 11
}
```

I want to Update  "nestedUpdateBehaviorAttribute":  as null

in which case DDB should be updated as 
```json
{
  "id": "ignoreNullFalse",
  "nestedRecord": {
    "id":"nestedId",
    "nestedUpdateBehaviorAttribute": null,
    "nestedCounter": 9999
  },
  "version": 12
}
```

also could you please update a test case for this scenario

Reply:
We do not handle updating record values to null for nested / or non-nested records as of today. 

```
@Test
    public void when_updatingNestedObjectWithSingleLevelWithNullOrEmpty_existingInformationIsPreservedAndUpdated_ignoreNulls() {

        NestedRecordWithUpdateBehavior nestedRecord = createNestedWithDefaults("id456", 5L);

        RecordWithUpdateBehaviors record = new RecordWithUpdateBehaviors();
        record.setId("id123");
        record.setCreatedOn(INSTANT_2);
        record.setNestedRecord(nestedRecord);

        mappedTable.putItem(record);

        NestedRecordWithUpdateBehavior updatedNestedRecord = new NestedRecordWithUpdateBehavior();
        long updatedNestedCounter = 10L;
        updatedNestedRecord.setNestedCounter(updatedNestedCounter);
        //updatedNestedRecord.setNestedUpdateBehaviorAttribute(null);

        RecordWithUpdateBehaviors update_record = new RecordWithUpdateBehaviors();
        update_record.setId("id123");
        update_record.setCreatedOn(null);
        update_record.setVersion(1L);

        update_record.setNestedRecord(updatedNestedRecord);

        mappedTable.updateItem(r -> r.item(update_record).ignoreNulls(true));

        RecordWithUpdateBehaviors persistedRecord = mappedTable.getItem(r -> r.key(k -> k.partitionValue("id123")));

        verifySingleLevelNestingTargetedUpdateBehavior(persistedRecord, updatedNestedCounter);
        assertThat(persistedRecord.getCreatedOn()).isNull();
```

The above test fails when we update out of our non-nested parent level attributes to null, and that is becasue of the way itemToMap models data. To change this behaviour we need to touch itemToMap. 

Diving deeper : we ignore attributes that are "null" and cannot differentiate between values set to null, and values that are null because they have not been updated in that operation. This is a bug that we had originally thought about, but it involves fundamental changes to itemToMap so we park it as an adjacent problem.  Refer to https://github.com/aws/aws-sdk-java-v2/blob/91c88084af5e0aeb779f00b68b014bdd6f9f8425/services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/mapper/StaticImmutableTableSchema.java#L520. We would need a way of differentiating between user set nulls and nulls set due to lack of updates


Im not exactly sure what you mean when we refer to an "empty" nested record. I can test it for cases when String attributes are updated to null